# Проект по SQL 

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

`Цель проекта`: проанализировать базу данных

`Данные`:
- **Таблица `books` - cодержит данные о книгах**
- **Таблица `publishers` - cодержит данные об издательствах**
- **Таблица `ratings`- cодержит данные о пользовательских оценках книг**
- **Таблица `reviews` - cодержит данные о пользовательских обзорах на книги**

In [1]:
!pip install sqlalchemy
!pip install pandas

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


<mark>Обзор данных:<mark>

**books**

In [3]:
query = ('''
        SELECT *
        FROM books
        LIMIT 10
        ''')
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

**publishers**

In [4]:
query = ('''
        SELECT *
        FROM publishers
        LIMIT 10
        ''')
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

**ratings**

In [5]:
query = ('''
        SELECT *
        FROM ratings
        LIMIT 10
        ''')
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

**reviews**

In [6]:
query = ('''
        SELECT *
        FROM reviews
        LIMIT 10
        ''')
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

## Задания

### 1. Посчитайте, сколько книг вышло после 1 января 2000 года

In [7]:
query = '''
        SELECT COUNT(book_id)
        FROM books
        WHERE publication_date > '2000-01-01'
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

#### После 1 января 2000 года вышло `819` книг

### 2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [8]:
query = '''
        SELECT b.book_id AS ID, b.title AS Title, 
        COUNT(DISTINCT rev.review_id) AS reviews, AVG(rat.rating) AS average_rating
        FROM books AS b 
        LEFT JOIN reviews AS rev 
        ON b.book_id = rev.book_id
        LEFT JOIN ratings AS rat 
        ON b.book_id = rat.book_id
        GROUP BY b.book_id, b.title
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

id                                              title  reviews  \
0       1                                       'Salem's Lot        2   
1       2                 1 000 Places to See Before You Die        1   
2       3  13 Little Blue Envelopes (Little Blue Envelope...        3   
3       4  1491: New Revelations of the Americas Before C...        2   
4       5                                               1776        4   
..    ...                                                ...      ...   
995   996           Wyrd Sisters (Discworld  #6; Witches #2)        3   
996   997                        Xenocide (Ender's Saga  #3)        3   
997   998                                    Year of Wonders        4   
998   999                        You Suck (A Love Story  #2)        2   
999  1000  Zen and the Art of Motorcycle Maintenance: An ...        4   

     average_rating  
0          3.666667  
1          2.500000  
2          4.666667  
3          4.500000  
4          4.000000  
..              ...  
995        3.666667  
996        3.400000  
997        3.200000  
998        4.500000  
999        3.833333  

[1000 rows x 4 columns]

#### Вывели книги, количество их ревью и средний рейтинг. 

### 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [9]:
query = ''' 
        SELECT pub.publisher, COUNT(b.book_id) as books_count 
        FROM books AS b
        LEFT JOIN publishers AS pub ON b.publisher_id = pub.publisher_id
        WHERE num_pages > 50
        GROUP BY publisher
        ORDER BY books_count DESC
        LIMIT 1            
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  books_count
0  Penguin Books           42

#### Издательство, которое выпустило наибольшее количество книг, страниц в которых больше 50, - `Penguin Books`. Оно выпустило `42` книги

### 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [10]:
query = '''
        SELECT
        au.author AS author,
        AVG(rat.rating) AS average_rating
        FROM ratings AS rat
        JOIN books AS b ON b.book_id = rat.book_id
        JOIN authors AS au ON b.author_id = au.author_id
        WHERE rat.book_id IN (
                               SELECT book_id
                               FROM ratings 
                               GROUP BY book_id
                               HAVING COUNT(book_id) >= 50)
        GROUP BY au.author_id
        ORDER BY AVG(rat.rating) DESC
        LIMIT 1'''
pd.io.sql.read_sql(sql=text(query), con = con)

author  average_rating
0  J.K. Rowling/Mary GrandPré        4.287097

#### Автором с самой высокой средней оценкой среди книг толще 50 страниц – J.K. Rowling/Mary GrandPré(последняя, кстати, иллюстратор книг по Гарри Поттеру). 

### 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [11]:
query = '''
        SELECT AVG(count_reviews)
        FROM 
             (SELECT COUNT(*) AS count_reviews
             FROM reviews
             WHERE username IN (SELECT username
                                FROM ratings
                                GROUP BY username
                                HAVING COUNT(*) > 50) 
             GROUP BY username) AS reviews
        '''
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.333333

#### В среднем пользователи оставляют `24` оценки

### Общий вывод

1. **После 1 января 2000 года было выпущено `819` книг**
2. **Penguin Books выпустило больше всего книг - `42` (книги, где более 50 страниц)**
3. **J.K. Rowling/Mary GrandPré автор с наивысшим рейтингом.** 
4. **Авторы, прочитавшие более 50 книг в среднем оставляют 24 оценки**